File: solutions/javascript/meetup/1/meetup.js
const WEEKDAYS = {
  'Sunday': 0, 'Monday': 1, 'Tuesday': 2, 'Wednesday': 3,
  'Thursday': 4, 'Friday': 5, 'Saturday': 6
};

export const meetup = (year, month, week, dayOfWeek) => {
  const targetDay = WEEKDAYS[dayOfWeek];
  let candidates = [];


  const date = new Date(year, month - 1, 1);
  while (date.getMonth() === month - 1) {
    if (date.getDay() === targetDay) {
      candidates.push(date.getDate());
    }
    date.setDate(date.getDate() + 1);
  }


  let day;
  switch (week) {
    case 'first':  day = candidates[0]; break;
    case 'second': day = candidates[1]; break;
    case 'third':  day = candidates[2]; break;
    case 'fourth': day = candidates[3]; break;
    case 'last':   day = candidates[candidates.length - 1]; break;
    case 'teenth': 
 
      day = candidates.find(d => d >= 13 && d <= 19);
      break;
  }

  return new Date(year, month - 1, day);
};

File: solutions/kotlin/meetup/1/src/main/kotlin/Meetup.kt
import java.time.DayOfWeek
import java.time.LocalDate

class Meetup(val month: Int, val year: Int) {

    fun day(dayOfWeek: DayOfWeek, schedule: MeetupSchedule): LocalDate {
        // Создаем начальную дату месяца
        val startDay = LocalDate.of(year, month, 1)
        // Находим количество дней в этом месяце
        val lengthOfMonth = startDay.lengthOfMonth()
        
        // Создаем список всех дат месяца, которые совпадают с нужным днем недели
        val matchingDates = mutableListOf<LocalDate>()
        for (day in 1..lengthOfMonth) {
            val date = LocalDate.of(year, month, day)
            if (date.dayOfWeek == dayOfWeek) {
                matchingDates.add(date)
            }
        }

        // Выбираем дату в зависимости от правила schedule
        return when (schedule) {
            MeetupSchedule.FIRST  -> matchingDates[0]
            MeetupSchedule.SECOND -> matchingDates[1]
            MeetupSchedule.THIRD  -> matchingDates[2]
            MeetupSchedule.FOURTH -> matchingDates[3]
            MeetupSchedule.LAST   -> matchingDates.last()
            MeetupSchedule.TEENTH -> matchingDates.first { it.dayOfMonth in 13..19 }
        }
    }
}
